digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -70,17 +71,392 @@ class NodesAndJunctionsService() {
+    withDynSession {
+      time(logger, "Fetch nodes point templates") {
+        nodePointDAO.fetchTemplatesByBoundingBox(boundingRectangle)
       }
     }
   }
 
-  def getNodeById(nodeId: Long): Option[Node] = {
-    withDynSession{
-      time(logger, "Fetch node by ID") {
-        nodeDAO.fetchById(nodeId)
+  def getJunctionTemplatesByBoundingBox(boundingRectangle: BoundingRectangle): Seq[JunctionPoint] = {
+    withDynSession {
+      time(logger, "Fetch nodes point templates") {
+        junctionPointDAO.fetchTemplatesByBoundingBox(boundingRectangle)
       }
     }
   }
 
+  def removeObsoleteNodesAndJunctions(roadwayChanges: Seq[ProjectRoadwayChange], endDate: Option[DateTime], username: String = "-"): Unit = {
+    /*
+      // TODO Handle these rules regarding road numbers:
+      The node point is expired if it no longer has justification for the current network, that is, it is no longer a point of change of the road part,
+      the point of change of the road type and will no longer be linked to the junctions of the main roads (less than 20,000 road numbers and 40000-70000 road roads).
+
+      The node is expired if no node points or junctions are connected to it. If, after the changes, the node is joined only with ramps,
+      etc. (road number between 20000-40000), those junctions will become junction templates. The user must join junction templates to another existing node.
+    */
+    def obsoleteNodesAndJunctionsPointsOfTermination(terminatedRoadwayNumbers: Seq[Long]): (Seq[NodePoint], Seq[JunctionPoint]) = {
+      val obsoleteRoadwayPoints = roadwayPointDAO.fetchByRoadwayNumbers(terminatedRoadwayNumbers)
+      val obsoleteNodePoints = nodePointDAO.fetchByRoadwayPointIds(obsoleteRoadwayPoints.map(_.id))
+      val obsoleteJunctionPoints = junctionPointDAO.fetchByRoadwayPointIds(obsoleteRoadwayPoints.map(_.id))
+      (obsoleteNodePoints, obsoleteJunctionPoints)
+    }
+
+    def obsoleteNodesAndJunctionsPointsByRoadwayNumbers(affectedRoadwayNumbers: Seq[Long]): (Seq[NodePoint], Seq[JunctionPoint]) = {
+      val roadwayPoints = roadwayPointDAO.fetchByRoadwayNumbers(affectedRoadwayNumbers)
+      val nodePoints = nodePointDAO.fetchByRoadwayPointIds(roadwayPoints.map(_.id))
+      val junctionPoints = junctionPointDAO.fetchByRoadwayPointIds(roadwayPoints.map(_.id))
+
+      val obsoleteJunctions = junctionDAO.fetchObsoleteById(junctionPoints.map(_.junctionId).distinct)
+      val obsoleteJunctionPoints = junctionPointDAO.fetchJunctionPointsByJunctionIds(obsoleteJunctions.map(_.id))
+
+      val obsoleteNodes = nodeDAO.fetchObsoleteById(nodePoints.map(_.nodeId.get).distinct)
+      val obsoleteNodePoints = nodePointDAO.fetchNodePointsByNodeId(obsoleteNodes.map(_.id))
+
+      (obsoleteNodePoints, obsoleteJunctionPoints)
+    }
+
+    def obsoleteNodesAndJunctionsPointsOfNew(newRoadwayNumbers: Seq[Long]): (Seq[NodePoint], Seq[JunctionPoint]) = {
```

Comment:
This method should be shorter, some sub-methods can be created

Reply:
Done. Commiting...